Producer/src/GenParticlesFiller.cc

```diff
@@ -123,10 +123,12 @@ struct PNodeWithPtr : public PNode {
           // found a matching candidate, kick it out
           mother->daughters[iD] = this;
           replacedCandPtr = static_cast<PNodeWithPtr*>(d)->candPtr;
-          // this has to be a GenParticlePtr, but just to check
-          GenParticlePtr genP(replacedCandPtr);
-          if (replacedCandPtr.isNonnull())
-            statusBits = genP->statusFlags().flags_;
+
+          if (dynamic_cast<reco::GenParticle const*>(&*replacedCandPtr)) {
```

Comment:
Sorry, I didn't test this because I thought edm::Ptr is supposed to do dynamic_cast internally. If you need to cast by hand, there's no need to create a GenParticlePtr:
```auto* genP(dynamic_cast<reco::GenParticle const*>(&*replacedCandPtr));
if (genP != nullptr)
  statusBits = genP->statusFlags().flags_;```
will be cleaner.

Reply:
It doesn't come out as a null pointer either though. That was the first thing I tried. The compiler doesn't complain since they both inherit from reco::Candidate. I guess it thinks it's fine at runtime too up until it tries to access a member that doesn't exist (not that the pointer isn't set).

Reply:
Hi Dan, that's a different problem then. Dynamic cast is used exactly for the purpose of identifying the correct type at runtime. If the result of dynamic cast is non-null, the pointer is of a valid type as far as the program is concerned. Does this happen in a specific event?

Reply:
Sorry, I might not have been clear. The edm::Ptr genP was not null, even when dynamic_cast is false. So I'm guessing the constructor doesn't actually do a dynamic_cast. The dynamic_cast is false for maybe two to four particles in 100 events, so I haven't narrowed down what causes it.

Reply:
I see, then I was not clear in my original comment. What I meant was that you don't need to create `GenParticlePtr` and just work with `reco::GenParticle const*`:
```
replacedCandPtr = static_cast<PNodeWithPtr*>(d)->candPtr;
  
auto* genP(dynamic_cast<reco::GenParticle const*>(&*replacedCandPtr));
if (genP != nullptr)
  statusBits = genP->statusFlags().flags_;
```
And the comment is basically irrelevant (you might gain a couple of milliseconds ;)

Reply:
Ah, I understand now. I didn't realize dynamic_cast could be used that way.